elastic/kibana
# [Security Solution][Detections] Adds confirmation modal with docs link when updating rules if V1/V2 jobs are installed

**Resolves:** https://github.com/elastic/kibana/issues/128121
**Related to:** https://github.com/elastic/detection-rules/pull/1846, https://github.com/elastic/kibana/pull/133845

## Summary

This PR adds a confirmation modal that we show when a user tries to load/upgrade their detection rules and has one of the affected ML jobs installed (doesn't have to be enabled) that are being replaced by the new V3 jobs.

NOTE: rules started pointing to the new V3 jobs in https://github.com/elastic/detection-rules/pull/1846 and starting from https://github.com/elastic/kibana/pull/133845 in Kibana.

There are three buttons that users can use to upgrade their rules. If they have one of the affected jobs installed AND have ML privileges, they should expect to see the modal when clicking any of those buttons. If they have one of the affected jobs installed AND DON'T have ML privileges, it is not expected that they would see the confirmation modal. And finally, if they don't have any of the affected jobs installed they shouldn't see the modal.

You can find these three buttons in:

- "Empty screen" of the Rules table when no rules exist. See the `PrePackagedRulesPromptComponent`
- The main header on the Rule Management page. See the `RulePageComponent`
- The "new prebuilt rules available" callout. See the `UpdatePrePackagedRulesCallOutComponent` (which uses a function wrapped by the `RulePageComponent`).

Here's the list of affected job IDs: https://github.com/elastic/kibana/blob/eda41a2b5e829b57f1de8a2a2c45377c45aa6df7/x-pack/plugins/security_solution/public/detections/components/callouts/ml_job_compatibility_callout/affected_job_ids.ts#L23-L72

## Screenshots

##### Modal with initial copy:
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172502879-ddce03d7-930e-436c-be12-8a9e8f04ecbe.png" />
</p>

##### Modal with overflow:
<p align="center">
  <img width="500" src="https://user-images.githubusercontent.com/2946766/172664681-8f5d99dc-8533-4d8d-b664-c6f37c14c9ed.png" />
</p>

## How to test

Since this modal will only show if you have one of the affected v1/v2 jobs, and these jobs were removed in 8.3, the easiest way to exercise this codepath is to just create a custom ML job using one of the affected job names, e.g. `v2_rare_process_by_host_linux_ecs`. Otherwise you would need to load an 8.2 instance, install the jobs, then go through an upgrade to 8.3 (which the ML folks have already tested, so no need to verify that part of the flow again).

Since we can only add this check client-side if the user has ML privileges, when testing with a user _without_ ML privileges, you should expect _not to see_ a confirmation modal.

<details><summary>Create Custom ML Job Instructions</summary>

* Go to ML App and click Jobs/Create
<img width="1402" alt="Anomaly_Detection_Jobs_-_Machine_Learning_-_Elastic" src="https://user-images.githubusercontent.com/2946766/172655095-00f6120c-3085-41e0-86f0-aead499c2da7.png">

* Select Security Solution Data View
<img width="1402" alt="Anomaly_Detection_Jobs_-_Machine_Learning_-_Elastic" src="https://user-images.githubusercontent.com/2946766/172655192-af2f775b-6d77-4470-952c-e9c4f6bb6c20.png">

* Select Single Metric
<img width="1394" alt="Anomaly_Detection_Jobs_-_Machine_Learning_-_Elastic" src="https://user-images.githubusercontent.com/2946766/172655285-d733cf03-62d3-4cf9-8d7e-20d60e1db990.png">

Click next
Pick a field
Click next
Input `v2_rare_process_by_host_linux_ecs` as the job id
Then `security` for the group
Click next
Click next
Click create Job
Go back to the Security  Rules Management page

</details>



### Checklist

Delete any items that are not applicable to this PR.

  * Collaborating with @elastic/security-docs folks in dedicated docs issue: https://github.com/elastic/security-docs/issues/1748
